Charts: Enhance BarChart component to support horizontal orientation

chihsuan opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Related to CHARTS-7

## Proposed changes:
<!--- Explain what functional changes your PR includes -->

This PR enhances the BarChart component to support horizontal orientation by adding `orientation` prop and updating the rendering logic accordingly.

<img width="874" alt="Screenshot 2025-06-03 at 15 59 25" src="https://github.com/user-attachments/assets/7e9267e0-e9ec-4d29-bdb1-bc6440fba11c" />

The method is simple: enabling horizontal orientation swaps the x-axis and y-axis configurations, along with a few other adjustments.

- Updated `useChartMargin` to calculate margins based on the chart orientation.
- Adjusted rendering logic for bars and tooltips to align with the selected orientation.
- Added a new story for `HorizontalBarChart`

Note that this PR does not fully address the CHARTS-7 issue, as that task requires additional customization. As mentioned in the discussion, I don't think the existing bar chart component should be extended to support this.

This PR focuses on adding support for horizontal bar charts. In a follow-up PR, I will either create a new chart or demonstrate how to build the specific horizontal bar chart required for WA.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

1 Go to charts folder
```sh
cd projects/js-packages/charts/
```

2. Run Storybook app instance
```sh
npm run storybook
```

3. http://localhost:50240/?path=/story/js-packages-charts-types-bar-chart--horizontal-bar-chart
4. Confirm the horizontal bar chart is displayed correctly.
5. Confirm the other bar chart stories are still displayed correctly

chihsuan committed:
Enhance BarChart component to support horizontal orientation and improve axis tick formatting

- Added `orientation` prop to `BarChart` for toggling between horizontal and vertical layouts.
- Updated axis tick formatting logic to accommodate the new orientation, ensuring correct display of tick labels.
- Refactored `useChartMargin` to calculate margins based on the chart orientation.
- Adjusted rendering logic for bars and tooltips to align with the selected orientation.

chihsuan committed:
Refine BarChart grid visibility logic and add horizontal orientation story

- Updated `gridVisibility` handling in the `BarChart` component to default based on the selected orientation.
- Enhanced type documentation for `gridVisibility` to clarify default behavior.
- Added a new story for `HorizontalBarChart` demonstrating the component with horizontal orientation and grid visibility set to 'none'.

chihsuan committed:
changelog

chihsuan committed:
Fix type

kangzj commented on projects/js-packages/charts/src/components/bar-chart/bar-chart.tsx:
```diff
@@ -70,43 +71,60 @@ const BarChart: FC< BarChartProps > = ( {
 					orientation: 'bottom' as const,
 					numTicks: 4,
-					tickFormat: getDefaultXTickFormat( data?.[ 0 ]?.data ),
+					tickFormat: horizontal ? defaultYTickFormat : defaultXTickFormat,
```
It feels it might make sense centralize those options to a hook, so that we don't have to use ternary operator all over the place.

chihsuan replied:
Sounds good! I've added a new hook to centralize those options. 5e2592f4c4f9a6e6f3274a8f39f4fd87edb0333e

chihsuan committed:
Refactor BarChart component and introduce useBarChartOptions hook

- Removed redundant code and improved readability in the BarChart component.
- Introduced a new `useBarChartOptions` hook to encapsulate chart options logic, enhancing maintainability.
- Updated axis and scale configurations based on chart orientation.
- Simplified tooltip label formatting and accessors for better clarity and performance.

chihsuan committed:
Reorder props

kangzj reviewed (dismissed):
It works as described and the other stories of bar charts are still looking good. Nice work!

Happy for the PR to merge now. I'm wondering if we should simplify the interface by renaming the tick format props to be more semantic—using `labelTickFormat` and `valueTickFormat` instead of tickFormats for x and y axis. This way, when someone changes the chart orientation, the tick formats would automatically apply to the correct axes without requiring additional prop changes.

---UPDATE---
Or we can keep the names for tickFormats, but we could use x always for labels and y always for values. Thoughts? 🤔

chihsuan committed:
Merge branch trunk into add/horizontal-bar-chart

chihsuan committed:
Refactor bar chart options to improve accessor and formatter handling

chihsuan commented:
> Happy for the PR to merge now. I'm wondering if we should simplify the interface by renaming the tick format props to be more semantic—using `labelTickFormat` and `valueTickFormat` instead of tickFormats for x and y axis. This way, when someone changes the chart orientation, the tick formats would automatically apply to the correct axes without requiring additional prop changes.

That's a great question! For now, I prefer to stick with `xTickFormat`/`yTickFormat`. Switching to more semantic prop names like `labelTickFormat` and `valueTickFormat` is a good suggestion, but since it would be a breaking change, I think we should consider it more carefully before making the switch. 🤔  

However, this give me an idea that we could improve the code in `useBarChartOptions` hook to make it more intuitive. I've updated the PR to change how we construct the default options in ff19fc5925d33ba4b1c76e2e39b688971420d161. Does this look good? Let me know if you have any thoughts.

> ---UPDATE--- Or we can keep the names for tickFormats, but we could use x always for labels and y always for values. Thoughts? 🤔

I think for new users, it might be unintuitive that x always means label, even if the label is on the y-axis in horizontal mode.

kangzj reviewed (approved):
:shipit: 

Great work 👍
